src/upgrade.py

```diff
@@ -0,0 +1,145 @@
+    with open("src/dependency.json") as dependency_file:
+        _deps = json.load(dependency_file)
+    return PostgreSQLDependencyModel(**_deps)
+
+
+class PostgreSQLUpgrade(DataUpgrade):
+    """PostgreSQL upgrade class."""
+
+    def __init__(self, charm, model: BaseModel, **kwargs) -> None:
+        """Initialize the class."""
+        super().__init__(charm, model, **kwargs)
+        self.charm = charm
+
+    @override
+    def build_upgrade_stack(self) -> List[int]:
+        """Builds ordered iterable of all application unit.ids to upgrade in.
+
+        Called by leader unit during :meth:`_on_pre_upgrade_check_action`.
+
+        Returns:
+            Iterable of integer unit.ids, LIFO ordered in upgrade order
+                i.e `[5, 2, 4, 1, 3]`, unit `3` upgrades first, `5` upgrades last
+        """
+        primary_unit_id = int(
+            self.charm._patroni.get_primary(unit_name_pattern=True).split("/")[1]
+        )
+        sync_standby_ids = [
+            int(unit.split("/")[1]) for unit in self.charm._patroni.get_sync_standby_names()
+        ]
+        unit_ids = [int(self.charm.unit.name.split("/")[1])] + [
+            int(unit.name.split("/")[1]) for unit in self.peer_relation.units
+        ]
+        # Sort the upgrade stack so replicas are upgraded first, then the sync-standbys
+        # at the primary is the last unit to be upgraded.
+        upgrade_stack = sorted(
+            unit_ids,
+            key=lambda x: 0 if x == primary_unit_id else 1 if x in sync_standby_ids else 2,
+        )
+        return upgrade_stack
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        """Log rollback instructions."""
+        logger.info(
+            "Run `juju refresh --revision <previous-revision> postgresql` to initiate the rollback"
+        )
+
+    @override
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        # Refresh the charmed PostgreSQL snap and restart the database.
+        self.charm.unit.status = MaintenanceStatus("refreshing the snap")
+        self.charm._install_snap_packages(packages=SNAP_PACKAGES, refresh=True)
+
+        if not self.charm._patroni.start_patroni():
+            logger.error("failed to start the database")
+            self.set_unit_failed()
+            return
+
+        self.charm._setup_exporter()
+        self.charm.backup.start_stop_pgbackrest_service()
+
+        # Wait until the database initialise.
+        self.charm.unit.status = WaitingStatus("waiting for database initialisation")
+        try:
+            for attempt in Retrying(stop=stop_after_attempt(6), wait=wait_fixed(10)):
+                with attempt:
+                    # Check if the member hasn't started or hasn't joined the cluster yet.
+                    if (
+                        not self.charm._patroni.member_started
+                        or self.charm.unit.name.replace("/", "-")
+                        not in self.charm._patroni.cluster_members
```

Comment:
same as here - https://github.com/canonical/postgresql-k8s-operator/pull/199#discussion_r1274848982

Reply:
Like above, I updated the wrong branch. This was updated on https://github.com/canonical/postgresql-operator/pull/187/commits/735b65412a97b432cd6eaa77768069023ad8cd4f in this branch/PR.